crossbar/worker/container.py

```diff
@@ -243,6 +243,47 @@ def panic(fail, msg):
                     session.disconnect()
                 session._swallow_error = panic
 
+                #   This is for RAW sockets where the onClose handler is in protocol session
+                #   rather than the protocol itself.
+                #
+                def close_wrapper(original_handler, was_clean):
+                    """
+                    Handle component shutdown (onClose)
+                    """
+                    return_value = original_handler(was_clean)
+                    #
+                    #   Make sure we're still open for business ...
+                    #
+                    if component_id not in self.components:
+                        self.log.warn('Component already closed ({})'.format(component_id))
+                        return return_value
+                    #
+                    #   Report the exit type
+                    #
+                    if was_clean:
+                        self.log.info('Component shutdown Ok ({})'.format(component_id))
+                    else:
+                        self.log.error('Component shutdown Errored ({})'.format(component_id))
+                    #
+                    #   Tidy up .. self.components is cleaned up in "stop_component"
+                    #
+                    component = self.components[component_id]
+                    self._publish_component_stop(component)
+                    component._stopped.callback(component.marshal())
+                    #
```

Comment:
Maybe this needs a better coment? What is this callback doing etc? (Is `_stopped` a `Deferred`.?) I guess what I'm confused about is where this data from the `.marshal()` end up going?

Reply:
Mm, this is pre-existing code I've just duplicated (see close_wrapper from on_connect_success) .. could probably be refactored to remove the duplication, and yes, there's lots of places where more comments would be good .. ;) .. let me see if I can refactor and comment .. but to answer your question; the problem I was seeing was this code not being called, then in stop_component a call to "yield _stopped" where stopped is "Deferred()" .. which means stop_component was effectively hanging waiting for a callback that would never come. All this is doing it signalling that the remaining component shutdown code is ok to continue .. I don't think the marshal data goes anywhere, it's just the code that was already there. Whereas it may be safe to remove, I'm not sufficiently familiar with / confident of the unit tests to remove stuff because it looks obsolete to me ..